distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Encoding.enso

```diff
@@ -48,19 +49,49 @@ type Encoding
             charset = Charset.forName name
             Encoding.Value charset.name
 
+    ## ICON convert
+       A default encoding that will try to guess the encoding based on some
+       heuristics.
+
+       If a BOM is present, it will be used to determine UTF-8 or UTF-16
+       encoding. Otherwise, the input is decoded using UTF-8 unless it contains
+       invalid UTF-8 sequences, in which case Windows-1252 is used as a fallback.
+
+       When used for encoding, it will always encode using UTF-8.
+       This encoding cannot be passed to some functions that require a Java
+       Charset.
+    Default
+
     ## PRIVATE
-       Create a new Encoding object.
+       An encoding identified by its Java name.
 
        Arguments:
        - character_set: java.nio.charset name.
-    Value (character_set:Text)
+    Value (internal_character_set:Text)
+
+    ## PRIVATE
+       Convert an Encoding to it's corresponding Java Charset, if applicable.
+       This method should be used in places not aware of special logic for the
+       Default encoding. In such places, usage of Default encoding will be forbidden.
+    to_java_charset self -> Charset ! Illegal_Argument =
+        self.to_java_charset_or_null.if_nothing <|
+            Error.throw (Illegal_Argument.Error "The Default encoding cannot be converted to a Java Charset. Please select a specific encoding for this operation.")
```

Comment:
I don't think so, I did this deliberately.

My idea is that for Write operations we should use the `Encoding.utf_8` as the default _explicitly_ OR in places where read/write is both possible and `Encoding.Default` is used, our write logic should have logic to consciously replace `Encoding.Default` with something else in Write mode.

This is nicely demonstrated in commit 5656d6a30e6d331f47842e0963714fe2998e9853

If I had this method select UTF-8 automatically, I would not have noticed that I'm lacking special logic for Delimited Write - it would just write as UTF-8. But that would be incorrect - e.g. when appending to a file that upon read is detected to be UTF-16 because it has a BOM - in such cases we want to also append in UTF-16, so we needed special logic changes.

Thus IMO it is better to keep as is - because whenever this error occurs it should be one of 3 scenarios:
1) we mistakenly put `Encoding.Default` as a default value for a write-only operation,
2) we forgot to have special handling for `Encoding.Default` in a read/write operation,
3) the user explicitly selected `Encoding.Default` for a write operation.

In case of (3) the error makes sense because the user should just select an encoding explicitly.

In cases (1) and (2) it also means that we actually forgot to implement something and we should fix the code - otherwise we could get invalid behaviour and corrupted files - e.g. with mixed UTF-8 and UTF-16.

Reply:
Ah I think I somehow missed the 'attach a warning' part.

I still think the error will be more noticeable. But OTOH the warning will allow the user to proceed so maybe that is the right call indeed.

Amended.